gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.util.concurrent;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.Iterables;
+
+import lombok.Synchronized;
+
+/**
+ * Base class for a task scheduler that can run {@link ScheduledTask}s periodically. Subclasses can implement
+ * {@link #startImpl(Optional)} and {@link #closeImpl()}.
+ */
+public abstract class TaskScheduler<K, T extends ScheduledTask<K>> implements Closeable {
```

Comment:
`start` and `close` are thread-safe because they are synchronized.  `schedule` is thread-safe because the underlying `get` call is thread-safe.